src/topology/homotopy.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shing Tak Lam
+-/
+
+import topology.unit_interval
+import topology.algebra.ordered.proj_Icc
+import topology.continuous_function.basic
+
+/-!
+# Homotopy between functions
+
+In this file, we define a `homotopy` between two functions `f₀` and `f₁`.
+
+## Definitions
+
+* `homotopy f₀ f₁` is the type of homotopies between `f₀` and `f₁`
+* `homotopy.refl f₀` is the constant homotopy between `f₀` and `f₀`
+* `homotopy.symm f₀ f₁` is a `homotopy f₁ f₀` defined by reversing the homotopy
+* `homotopy.trans F G`, where `F : homotopy f₀ f₁`, `G : homotopy f₁ f₂` is a `homotopy f₀ f₂`
+  defined by putting the first homotopy on `[0, 1/2]` and the second on `[1/2, 1]`.
+-/
+
+noncomputable theory
+
+universes u v
+
+variables {X : Type u} {Y : Type v} [topological_space X] [topological_space Y]
+
+open_locale unit_interval
+
+/--
+The type of homotopies between two functions.
+-/
+
+structure homotopy (f₀ f₁ : X → Y) extends C(X × I, Y) :=
+(to_fun_zero : ∀ x, to_fun (x, 0) = f₀ x)
+(to_fun_one : ∀ x, to_fun (x, 1) = f₁ x)
+
+namespace homotopy
+
+section
+
+variables {f₀ f₁ : X → Y}
+
+instance : has_coe_t (homotopy f₀ f₁) C(X × I, Y) := ⟨homotopy.to_continuous_map⟩
```

Comment:
This should have a simp lemma too, either:

```lean
@[simp] lemma to_continuous_map_eq_coe (F : homotopy f₀ f₁) : F.to_continuous_map = ↑F := rfl
```
or
```lean
@[simp] lemma coe_eq_to_continuous_map (F : homotopy f₀ f₁) : ↑F = F.to_continuous_map := rfl
```

Reply:
Yeah, I'm currently leaning towards the coercion being a bad idea